tensorflow/compiler/jit/xla_gpu_device.cc

```diff
@@ -51,14 +51,18 @@ Status XlaGpuDeviceFactory::CreateDevices(const SessionOptions& options,
   return Status::OK();
 }
 
+#if GOOGLE_CUDA
```

Comment:
Now I think about this a bit more closely, there's a problem here.

There are two separate things going on here:
a) whether you have the CUDA libraries and enabled CUDA at "configure" time.
b) whether you are building with a CUDA compiler (by passing --config=cuda to Bazel).

For most parts of TensorFlow, these are identical. However for XLA, we need only (a) but not (b) --- XLA generates its own GPU code without needing nvcc.

The PR as written compiles the GPU registrations only in case (b). However we do support (and indeed have continuous builds that test) that you can run the XLA tests if (a) is true but (b) is not.

The correct way to do this would be to use the "if_cuda_is_configured()" bazel macro to conditionally compile this code.

You can see an example of how this works here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/BUILD#L41
The if_cuda_is_configured macro expands to the empty list if (a) is false, or its argument, if (a) is true. So we can use it to add or remove files from the build if CUDA is configured.

For xla_gpu_device.cc, you can probably achieve this by wrapping the two GPU dependencies of this build target in a "if_cuda_is_configured" block. That target is the one linked in by the Tensorflow Python runtime if you enable XLA.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/jit/BUILD#L35
(You'll need to add if_cuda_is_configured to the "load" statements, as in the stream_executor BUILD file.)

For xla_op_registry.cc, you'll need to move the REGISTER_XLA_BACKEND and GpuOpFilter into a separate file (say xla_gpu_backend.cc) and compile that file conditionally as well.

Make sense?